### PR TITLE
Added an implantation module

### LINKED ISSUE #29: Add an implantation module
Similar to the `extraction.py` file, an `implantation.py` file should be created and an `implant` function should be defined. The `implant` function takes a COCO JSON/dict and creates a `.ts.png` file. The inverse of the extraction. Unit tests would be appreciated for this as well.

### PR DESCRIPTION
**Pull request by gabb222**
Fixes #29 

The implantation module does not currently work, but it is uploaded to save progress and potentially get a second look at how to make it work.

### CODE REVIEW

File: tests/assets/coco_data.json
Comment (volks73) [inline tests/assets/coco_data.json:1]: The content of this file is not valid JSON. Only line 24 is valid JSON.

File: tspng/implantation.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        Raises:
+                TypeError: If not a BytesIO object, file, list of files, or folder
+    '''
+    #call appropriate function
+    if type(file) == str and os.path.isfile(file):
+        return implant_to_file(file, image, MIME_TYPE)
+    else:
+        raise TypeError(f"{file} is not a BytesIO object, file, list of files, or folder.")
+
+def implant_to_file(path: str, image: str, mime_type: str=MIME_TYPE) -> Dict:
```
Comment (volks73) [inline tspng/implantation.py]: This should be renamed to `implant_from_file`. Implanting is the inverse of extracting. So, while we would extract to a file, we would implant from a file.

Other "froms" would include a string, JSON, and eventually another tspng.

File: tspng/implantation.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        Returns:
+                (dict): Dictionary containing file metadata
+
+        Raises:
+                TypeError: If not a BytesIO object, file, list of files, or folder
+    '''
+    #call appropriate function
+    if type(file) == str and os.path.isfile(file):
+        return implant_to_file(file, image, MIME_TYPE)
```
Comment (volks73) [inline tspng/implantation.py]: The `MIME_TYPE` should be lower case, i.e. it should be argument variable for the function and not the constant.

File: tests/assets/coco_data.json
Comment (volks73) [inline tests/assets/coco_data.json:1]: The contents of the file were not valid JSON. Only line 24 was JSON.

File: tspng/test_implantation.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+from io import BytesIO
+from tspng.implantation import implant,implant_to_file
+from tspng.extraction import extract
+
+def test_implant():
+    '''
+    Tests the dictionary keys from an example TSPNG file path.
+    '''
+    #test file
+    implant('tests/assets/coco_data.txt','tests/assets/empty.png')
```
Comment (volks73) [inline tspng/test_implantation.py:16]: The file is `tests/assets/coco_data.json`.

File: tspng/implantation.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+        Returns:
+                (dict): Dictionary containing file metadata
+
+        Raises:
+                TypeError: If not a BytesIO object, file, list of files, or folder
+    """
+    # call appropriate function
+    if type(file) == str and os.path.isfile(file):
+        return implant_to_file(file, image, mime_type)
```
Comment (volks73) [inline tspng/implantation.py:43]: Changed the case, so we are good.

File: tspng/implantation.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+# import statements
+import os
+
+from PIL import Image
+from PIL.PngImagePlugin import PngInfo
+from tspng import MIME_TYPE
+from typing import Dict
+
+
+def _implant_data(file: str, image: str, mime_type: str = MIME_TYPE) -> Dict:
```
Comment (volks73) [inline tspng/implantation.py:10]: Eventually, the `file` argument should be changed to `data: str` and it should be a valid JSON string. Then, the `implant_to_file` renamed to `implement_from_file`, which would read the file, validate as JSON, and then pass to this function as the `data: str`.

**Review by volks73 — CHANGES_REQUESTED** [review]
I have pushed some immediate changes as I was working through the PR.

I have made some comments that hopefully will help with direction. Great start!

**Comment by gabb222** [comment]
Updated the branch with a working implantation module with unit tests. @volks73 ready for review.

File: tspng/implantation.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import os
+import json
+
+from PIL import Image
+from PIL.PngImagePlugin import PngInfo
+from tspng import MIME_TYPE
+from typing import Dict
+
+
+def _implant_data(data: str, image: str, mime_type: str = MIME_TYPE) -> Dict:
```
[the remark this answers is not in the document]
Comment (volks73) [reply tspng/implantation.py:11]: Looking at the Pillow documentation, the `save` function returns `None`. The type annotation for the return of the function can be dropped complete, `-> Dict` deleted.

File: tspng/test_implantation.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+from tspng.implantation import implant,implant_from_file
+from tspng.extraction import extract
+
+def test_implant():
+    '''
+    Tests the dictionary keys from an example TSPNG file path.
+    '''
+    #test file
+    implant('tests/assets/coco_data.json','tests/assets/empty.png')
+    test_data = extract('tests/assets/empty.png_MOD')
```
Comment (volks73) [inline tspng/test_implantation.py:17]: Once the `.ts.png` file extension is applied, this needs to be updated and then the `tests/assets/empty.ts.png` file will need to be deleted after extracting the data but before the assertion so if any of the assertions that follow do not cause the file to linger.

File: tspng/implantation.py (hunk 6 above)
Comment (volks73) [inline tspng/implantation.py:11]: To make this more easily testable, a `dst` argument, short for `destination`, should be added and implemented. It would be `Optional[str]` type and default to `None`. If the value is `None`, then the image file is created in the same location as the `image` file. If the value is a path as a string to a directory that already exists, then the `.ts.png` image will be saved to the directory with the file stem being the same as the `image` file. If the value is a path as a string and it does not exist, then the assumption is that it should be the output file name.

With this added, the [`tmp_path`](https://docs.pytest.org/en/6.2.x/tmpdir.html) fixture can be used in tests without having to worry about cleaning up after tests.

File: tspng/test_implantation.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+def test_implant_from_file():
+    '''
+    Tests the dictionary keys from an example TSPNG file path.
+    '''
+    implant_from_file('tests/assets/coco_data.json','tests/assets/empty.png')
+    #test_data = extract('tests/assets/coco_data.txt+tests/assets/empty.png')
+    test_data = extract('tests/assets/empty.png_MOD')
+    assert list(test_data.keys())==['info','licenses','images','annotations','models','categories']
+
+def test_implant_from_file_fails():
```
Comment (volks73) [inline tspng/test_implantation.py:33]: This test is not quite right. It probably should be something like:

```python
def test_implant_from_file_fails_with_folder():
    with pytest.raises(Exception):
        implant_from_file("tests/assets", "tests/assets/empty.png")


def test_implant_from_file_fails_existence():
    with pytest.raises(Exception):
        implant_from_file("tests/assets/no_such_file.json", "tests/assets/empty.png")
```

File: tspng/test_implantation.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    Tests the dictionary keys from an example TSPNG file path.
+    '''
+    #test file
+    implant('tests/assets/coco_data.json','tests/assets/empty.png')
+    test_data = extract('tests/assets/empty.png_MOD')
+    assert list(test_data.keys())==['info','licenses','images','annotations','models','categories']
+
+def test_implant_fails():
+    with pytest.raises(TypeError):
+        implant('Test for failure')
```
Comment (volks73) [inline tspng/test_implantation.py:22]: The invocation of the `implant` function is missing the second required argument, the path to the image file.

File: tests/assets/coco_data.json
Comment (volks73) [inline tests/assets/coco_data.json:1]: For the first implementation of this feature, any kind of JSON data can be added, but in truth, for a `.ts.png`, there is a specific JSON schema used. This is just something to keep in mind, but does not need to be address in this PR.

**Review by volks73 — CHANGES_REQUESTED** [review]
Looking good, but some additional changes are needed.

**Comment by volks73** [comment]
I cleaned up the imports a little.

File: tspng/implantation.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    # open image
+    target_im = Image.open(image)
+    if target_im.format != "PNG":
+        raise Exception("Image is not a PNG.")
+    # implant data
+    metadata = PngInfo()
+    metadata.add_text(mime_type,data)
+    # save file with implanted data
+    mod_file = target_im.save(image+'_MOD', format="PNG", pnginfo=metadata)
```
[the remark this answers is not in the document]
Comment (gabb222) [reply tspng/implantation.py:20]: For now, I've implemented a roundabout way and can change this later on.

File: tests/assets/coco_data.json
Comment (gabb222) [reply tests/assets/coco_data.json:1]: I did not implement this but can address it in another PR.

**Comment by gabb222** [comment]
@volks73 Ready for another review

File: tspng/implantation.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+def _implant_data(data: str, image: str, mime_type: str = MIME_TYPE, dst: str = None):
+    # open image
+    target_im = Image.open(image)
+    if target_im.format != "PNG":
+        raise Exception("Image is not a PNG.")
+    # implant data
+    metadata = PngInfo()
+    metadata.add_text(mime_type,data)
+    # save file with implanted data
+    mod_file = target_im.save(image[:-4]+'.ts.png', format="PNG", pnginfo=metadata)
```
Comment (volks73) [inline tspng/implantation.py:20]: `image[:-4]` works, but all kinds of problems are going to occur. What if a TIFF file was passed with the extension, `.tiff`? Now, the `-4` is wrong.

I would use the [`splitext`](https://docs.python.org/3/library/os.path.html#os.path.splitext) function.

```python
base, _ = os.path.splitext(image)
base + ".ts.png"
```

File: tspng/implantation.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    # call appropriate function
+    if type(data) == str and os.path.isfile(data):
+        return implant_from_file(data, image, mime_type)
+    else:
+        raise TypeError(
+            f"{data} is not a file."
+        )
+
+
+def implant_from_file(path: str, image: str, mime_type: str=MIME_TYPE) -> Dict:
```
Comment (volks73) [inline tspng/implantation.py:49]: This has no return type. The `-> Dict` can be removed.

File: tspng/implantation.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import os
+import json
+
+from PIL import Image
+from PIL.PngImagePlugin import PngInfo
+from tspng import MIME_TYPE
+from typing import Dict
+
+
+def _implant_data(data: str, image: str, mime_type: str = MIME_TYPE, dst: str = None):
```
Comment (volks73) [inline tspng/implantation.py:11]: If the `dst` is not going to be used/implemented, then let's leave it out for now and add with a separate PR.

File: tspng/implantation.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    # open image
+    target_im = Image.open(image)
+    if target_im.format != "PNG":
```
Comment (volks73) [inline tspng/implantation.py:14]: This check for a PNG can be removed. The source image file could be any Pillow image format, PNG, TIFF, etc. and implanting will convert it to a TS png file (PNG). There is no restriction for the source image being PNG.

**Review by volks73 — CHANGES_REQUESTED** [review]
Some additional comments and requested changes, but once these are addressed, this is ready.

**Comment by gabb222** [comment]
These changes were made as requested, @volks73 here is the latest version!

**Review by volks73 — APPROVED** [review]
Looks good.
